# Imports

In [32]:
#Import libraries
import pandas as pd
import numpy as np
import os

In [33]:
# Import files
path = r'/Users/jorisjan/Documents/CareerFoundry/Instacart Basket Analysis'
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'departments_wrangled.csv'), index_col = False)

# Data Consistency Checks

## Creation of a small test dataframe

In [34]:
# Creation of the dataframe
df_test = pd.DataFrame()

# Creation of a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [35]:
df_test

mix
0     a
1     b
2     1
3  True

In [36]:
# check whether a dataframe contains any mixed-type columns
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [37]:
# check whether a dataframe contains any missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [38]:
# Create a subset of the dataframe containing only the missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [39]:
df_prods.shape

(49693, 5)

In [40]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]
df_prods_clean

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49677 rows x 5 columns]

In [41]:
# Check whether a dataframe contains any FULL duplicate records
df_dups = df_prods_clean[df_prods_clean.duplicated()]
# Display the DataFrame containing only rows that are duplicates
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [42]:
df_prods_clean.shape

(49677, 5)

In [43]:
# Create a new dataframe that doesn’t include identified duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

## 2. Run the df.describe() function on df_ords dataframe

In [44]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

The maximum order_number being 100 might be worth investigating to ensure it's not an outlier or data entry error

The maximum values in order_dow, order_hour_of_day, and days_since_prior_order are consistent with their respective scales and seem reasonable

The mean order_number is higher than the median, suggesting a right-skewed distribution where some users have very high order numbers

The standard deviation for order_number is almost as high as the mean, indicating a wide spread in the number of orders per user, this might suggest a diverse range of user engagement

## 3. Check for mixed-type data in df_ords dataframe.

In [45]:
# check whether a dataframe contains any mixed-type columns
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

As the code do not return any column name, the dataframe do not seems to contain columns with mixed-type data.

## 5. Run a check for missing values in df_ords dataframe.

In [46]:
# Check whether a dataframe contains any missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

The count of missing values for days_since_prior_order is 206,209 . It's common for the first order of a user to have a missing value in this field, as there's no prior order.

## 6. Address the missing values using an appropriate method.

In [47]:
# Create a new column with flags for missing values
df_ords['missing_flag'] = np.where(df_ords['days_since_prior_order'].isnull(), 'First Order', 'Subsequent Order')
# Display the DataFrame
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order      missing_flag  
0                        8                     NaN       First Order  
1                        7                    15.0  Subsequent Order  
2                       12                    21.0  Subsequent Order  
3                        7                    29.0  Subsequent Order  
4                       15                    28.0  Subsequent Order  
...                    ...                     ...               ...  
3421078                 18                    29.0  Subsequent Order  
3421079                 10                    30.0  Subsequent Order  
3421080                 12                    18.0  Subsequent Order  
3421081                 12                     7.0  Subsequent Order  
3421082                 14                    30.0  Subsequent Order  

[3421083 rows x 8 columns]

I used thif method to effectively identify first-time orders, to maintain the integrity of the data (instead of imputing values or removing these records, which could affect the analysis, I'm acknowledging the nature of these missing values, this way, I can conduct more nuanced analyses, such as comparing behavior between first-time and returning customers). Also I wanted to avoid misleading imputations as imputing missing values in this column could be misleading since the absence of a prior order is not the same as an average interval between orders. Flagging preserves the true nature of this data point.

## 7. Run a check for duplicate values in df_ords data.

In [48]:
# Check whether a dataframe contains any FULL duplicate records
df_ords_dups = df_ords[df_ords.duplicated()]
# Display the DataFrame containing only rows that are duplicates
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, missing_flag]
Index: []

As the code do not return any other row than the header, the dataframe do not seems to contain full duplicated values.

## 9. Export final, cleaned df_prods and df_ords data as “.csv” files

In [50]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_cleaned.csv'))
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_cleaned.csv'))